### ISSUE TITLE
More flexible ExternalName

### ISSUE THREAD

**Issue by stoklund** [stoklund/cretonne#194]
Cretonne uses an `ExternalName` when it refers to functions or global variables outside the function it's currently compiling. The current design has a couple downsides:

- While object files use strings to identify symbols, other Cretonne users don't. When embedded as the JIT compiler for a virtual machine, `ExternalName` is used to encode WebAssembly function indexes and well-known runtime library functions using various binary encoding hacks.
- In the future, Cretonne will need to create references to well-known runtime library functions when it legalizes things like `idiv.i64` on 32-bit platforms. We don't want Cretonne to have to known about the embedder's crazy binary encoding hacks for symbol names.
- Currently, an `ExternalName` can't be part of an `InstructionData` variant, so it isn't possible to define a `external_addr` instruction which takes the address of an external function or global. Instead we have to store the `ExternalName` instances in preamble tables and separate `function_addr` and `globalsym_addr` instructions which really do the same thing. A smaller `ExternalData` type without a `Drop` impl is desirable.

## Proposed alternative

This `ExternalData` enum would solve the above problems:

```rust
#[derive(Clone, Copy)]
pub enum ExternalName {
    // A name in a user-defined symbol table.
    // Cretonne does not interpret these numbers in any way.
    User { pub namespace: u32, pub index: u32 }

    // A test case function name with up to 10 ascii characters.
    // This is not intended for use outside test cases.
    TestCase { length: u8, ascii: [u8; 10] }

    // A well-known runtime library function.
    // This is used for runtime library function calls that
    // Cretonne needs to understand about, such as routines
    // for performing 64-bit division on 32-bit platforms.
    LibCall(LibCall)
}

pub enum LibCall {
    F32Floor,
    F64Floor,
    I64Div,
    ...
}
```

The `User` variant is easier to use for embedders who don't have to invent binary string encoding schemes. Embedders generating object files will need to maintain an external symbol table containing linkage names for functions and global variables.

The `TestCase` variant is only for the convenience of writing test cases with meaningful function names. Names are limited to 10 ASCII characters to avoid the memory allocation and `Drop` impl of a `String` or `Vec<u8>` as we have now.

Finally, the `LibCall` variant can be created internally to Cretonne as part of the legalization phase. There's nothing wrong with using it in the input IR too.

## Text format

We need a textual representation of external names. Currently we have `%ascii_names` and `#01234abcd` hexadecimal names. I propose that we drop the hexadecimal name variants and use:

- Each `LibCall` variant has a simple `%ascii_name`.
- Other ASCII names become `TestCase` names.
- For the user names, we need a new notation. Something like `u3/123` could work.

When parsing an external name like `%f32_floor`, we use the `LibCall` variant if there is an exact match, and fall back to a `TestCase` name otherwise. Since the `TestCase` variants are only used for testing, there is no harm in limiting their namespace in text files.

cc @pchickey

**Comment by pchickey** [comment]
For `TestCase` names, I assume we want the restriction we currently have for ascii_names - alphanumeric and `_` only. Should I expose a constructor that rejects any names that don't meet that requirement, or should I make it do a hex translation of any invalid bytes?

**Comment by pchickey** [comment]
I think the username notation starting with `u` will make for an ambiguity in the lexer, where tokens starting with alphabetic chars are scanned into words. For now I'll use `$123/456` as the syntax so that there is no ambiguity. I'm totally happy to change it.

**Comment by stoklund** [comment]
I agree that `TestCase` names should be restricted to alphanumeric ASCII + `_`, which is already what the lexer requires for its `Token::Name` variant. Since test files is the only way we expect `TestCase` names to appear, I think it is fine to simply have a constructor that takes a `&str` (or even `&[u8]`) and just checks the length.

OTOH, I definitely don't want to invent syntax for unprintable test case names, so a check for ASCII in the constructor may make sense.

The lexer does recognize a number of so-called "numbered entities", see the `numbered_entity()` method. These have the form ASCII alpha + digits: `v12`, `ebb45`, etc. You could handle the `u123/456` syntax by lexing `u123`, `/`, and `456` as three separate tokens, where `u123` would be a new kind of "numbered entity" token. (Maybe `u123:456` is better, avoiding the introduction of the `/` token).

The text format so far has used this numbered entity pattern instead of sigils to distinguish different kinds of identifiers. Only names with an unbounded namespace uses sigils (`%` and `#`) to avoid clashing with instruction names, values, etc.

**Comment by pchickey** [comment]
Thanks, all of those ideas sound great to me. I hadn't seen the numbered entities, I was trying to mock it up to look like the hex or name tokens. I like the `u123:456` syntax, so I'll go ahead with that.

**Comment by pchickey** [comment]
Should be closed by #197
